### PR TITLE
Adapt helm-org to recent changes in org-get-outline-path

### PR DESCRIPTION
**Pull request by aroig**
The signature and behaviour of `org-get-outline-path` has changed in org mode's master branch, and was last touched by this [commit](http://orgmode.org/cgit.cgi/org-mode.git/commit/lisp/org.el?id=105a44669716abd50377fdcfd03a9a11dc388b16).

This PR adapts to the new behaviour, making use of the newly introduced cache to avoid backtracking. This addresses #1421 for org-mode's current master only. I'm not sure how to be compatible with both new and old `org-get-outline-path` at the same time. Should we use `condition-case` to handle the error on wrong signature? @TheBB feel free to finish the patch! ;)

### CODE REVIEW

**Comment by TheBB** [comment]
Hey thanks for this, I was too busy and was starting to feel bad about my promise.

I would assume condition-case is the way to go. Something like this

```elisp
(condition-case nil
    ;; call old org function
  (wrong-number-of-arguments
    ;; call new org function
    ))
```

**Comment by aroig** [comment]
I force-pushed the `condition-case` thing. Now the PR is complete, and should work either with org master branch or older stable versions.

I have only tested in with org from master branch. Please make sure it works with older org-mode before merging.

File: helm-org.el [hunk 1]
```diff
@@ -189,10 +191,16 @@ Note this have no effect in `helm-org-in-buffer-headings'."
...
+                                    ;; org-get-outline-path changed in signature and behaviour since org's
+                                    ;; commit 105a4466971. Let's fall-back to the new version in case
+                                    ;; of wrong-number-of-arguments error.
+                                    (condition-case nil
+                                        (append (apply #'org-get-outline-path
+                                                       (unless parents
+                                                         (list t level heading)))
+                                                (list heading))
+                                      (wrong-number-of-arguments
+                                       (apply #'org-get-outline-path (list t t))))
```
Comment (TheBB) [inline helm-org.el]: Actually there's no reason this can't be `(org-get-outline-path t t)`

**Comment by aroig** [comment]
I force-pushed again addressing a comment by @TheBB.The comment was about using `(org-get-outline-path t t)` instead of `(apply #'org-get-outline-path (list t t))`

**Comment by thierryvolpiatto** [comment]
The questions is how do we have the same behavior of  `(org-get-outline-path arg1 arg2 arg3)`
and  `(org-get-outline-path)` with the new version ?

IOW Are you sure:

`(org-get-outline-path t t)` == `(org-get-outline-path arg1 arg2 arg3)`

And I don't see in the patch correspondance for `(org-get-outline-path)` i.e called with no args.

**Comment by thierryvolpiatto** [comment]
So what is the same behavior of:

`(org-get-outline-path)` with new version ?

File: helm-org.el [hunk 2]
```diff
@@ -174,6 +174,8 @@ Note this have no effect in `helm-org-in-buffer-headings'."
         (save-restriction
           (widen)
           (unless parents (goto-char (point-min)))
+          ;; clear cache for new version of org-get-outline-path
+          (setq org-outline-path-cache nil)
```
Comment (thierryvolpiatto) [inline helm-org.el]: Please use `(and (boundp 'org-outline-path-cache) (setq ...`.

**Comment by thierryvolpiatto** [comment]
So as soon I am sure we have the same behavior as before with new org I will merge.

Thanks for your work.

**Comment by aroig** [comment]
It looks like this function has changed considerably. I'm not really sure how it worked before. What was the point in passing `level` and `heading` arguments? The new version removes the old three arguments, and adds two for totally unrelated things: including current headline in the output, and enabling a cache.

The patch I wrote is based on my understanding of the new function, and checking that `helm-org-in-buffer-headings` and `helm-org-parent-headings` work as expected. I may be missing something though.

**Comment by aroig** [comment]
Force-pushed with the `bound-p` check for `org-outline-path-cache`.

**Comment by aroig** [comment]
To be clear, as far as I understand, the new version called as `(org-get-outline-path t t)` returns a list containing the path to the current outline entry. The first `t` argument makes it include the current heading at the end, and the second argument enables using a cache in `org-outline-path-cache` which is supposed to improve performance on sequential reads of an entire file.

Does helm-org require a different behaviour from this function?

**Comment by thierryvolpiatto** [comment]
Abdó Roig-Maranges <notifications@github.com> writes:

> It looks like this function has changed considerably.

Yes it is why I am asking about the new behavior.

> The patch I wrote is based on my understanding of the new function,
> and checking that helm-org-in-buffer-headings and
> helm-org-parent-headings work as expected.

I guess that if you tried the both and they are working as expected it
is all fine and we can merge.

I can't tell you more what the old version does exactly, I just made a
rewrite of the function based on the knowledge of original author.
My org knowledge is limited.

> I may be missing something though.

I have not the last org installed, for my basic org usage the default version
is enough, so I didn't check how works the new version.

Let's merge your changes and we will see what happen (crossing fingers
:-).

Thanks.

-- 
Thierry

**Comment by thierryvolpiatto** [comment]
Merged thanks.

**Comment by StreakyCobra** [comment]
Seems to work as expected here :+1: Thanks.

**Pull request merged**